// microframework-w3tec/MicroframeworkSettings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Bootstrap settings used across all loader.
 * Used to register shutdown handlers - special functions that will be executed when framework is shutdown.
 * Also can be used to store and share data across modules.
 */
var MicroframeworkSettings = /** @class */ (function () {
    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------
    function MicroframeworkSettings(config) {
        this.config = config;
        // -------------------------------------------------------------------------
        // Private Properties
        // -------------------------------------------------------------------------
        /**
         * Data which can be used to share data across modules.
         */
        this.data = {};
        /**
         * Special functions that will be executed when framework is shutdown.
         */
        this.shutdownHandlers = [];
    }
    // -------------------------------------------------------------------------
    // Public Methods
    // -------------------------------------------------------------------------
    /**
     * Sets data that can be used across other modules.
     */
    MicroframeworkSettings.prototype.setData = function (key, value) {
        this.data[key] = value;
        return this;
    };
    /**
     * Gets previously set data by this or before executed module.
     */
    MicroframeworkSettings.prototype.getData = function (key) {
        return this.data[key];
    };
    /**
     * Adds a shutdown handler - special function that will be executed when framework is shutdown.
     */
    MicroframeworkSettings.prototype.onShutdown = function (handler) {
        this.shutdownHandlers.push(handler);
        return this;
    };
    /**
     * Gets all shutdown handlers - special functions that will be executed when framework is shutdown.
     */
    MicroframeworkSettings.prototype.getShutdownHandlers = function () {
        return this.shutdownHandlers.map(function (handlers) { return handlers; });
    };
    return MicroframeworkSettings;
}());
exports.MicroframeworkSettings = MicroframeworkSettings;
